src/index/Vocabulary.cpp

```diff
@@ -1,25 +1,93 @@
-// Copyright 2011, University of Freiburg,
+// Copyright 2024, University of Freiburg,
 // Chair of Algorithms and Data Structures.
-// Authors: Björn Buchhold <buchholb>,
-//          Johannes Kalmbach<joka921> (johannes.kalmbach@gmail.com)
+// Authors: Björn Buchhold <buchhold@gmail.com>
+//          Johannes Kalmbach <kalmbach@cs.uni-freiburg.de>
+//          Hannah Bast <bast@cs.uni-freiburg.de>
 
-#include "./Vocabulary.h"
+#include "index/Vocabulary.h"
 
 #include <fstream>
 #include <iostream>
 
-#include "../parser/RdfEscaping.h"
-#include "../parser/Tokenizer.h"
-#include "../util/BatchedPipeline.h"
-#include "../util/File.h"
-#include "../util/HashMap.h"
-#include "../util/HashSet.h"
-#include "../util/Serializer/FileSerializer.h"
-#include "../util/json.h"
-#include "./ConstantsIndexBuilding.h"
+#include "index/ConstantsIndexBuilding.h"
+#include "parser/RdfEscaping.h"
+#include "parser/Tokenizer.h"
+#include "util/BatchedPipeline.h"
+#include "util/File.h"
+#include "util/HashMap.h"
+#include "util/HashSet.h"
+#include "util/Serializer/FileSerializer.h"
+#include "util/json.h"
 
 using std::string;
 
+// ____________________________________________________________________________
+template <typename StringType, typename ComparatorType, typename IndexT>
+void Vocabulary<StringType, ComparatorType, IndexT>::setRangesForPrefix(
+    PrefixRanges& ranges, std::string_view prefix) {
+  // NOTE: If a prefix is not present in a vocabulary, `first` and `last`of
+  // the range returned by `prefix_range` will both be equal.
+  auto rangeInternal = internalVocabulary_.prefix_range(prefix);
+  auto rangeExternal = externalVocabulary_.prefix_range(prefix);
+  ranges.rangeInternalBegin_ = IndexType::make(rangeInternal.first);
+  ranges.rangeInternalEnd_ = IndexType::make(rangeInternal.second);
+  auto offset = internalVocabulary_.size();
+  ranges.rangeExternalBegin_ = IndexType::make(rangeExternal.first + offset);
+  ranges.rangeExternalEnd_ = IndexType::make(rangeExternal.second + offset);
+}
```

Comment:
This function should only take the `prefix` and return the `PrefixRange`.

Reply:
I also thought about that as an alternative. Can you briefly explain why you think it is better to return the object (and then assign it when invoking the function) instead of setting it? Simply because that makes the function (slightly) more versatile?

Reply:
The `ranges` argument is a pure out argument as it is only ever written to, never read.
By default those should be returned (makes the interface clearer and easier to reason about,  we have return value optimization, etc.).
The only reason to not do something like this, is if the empty object is complicated to construct and we want to delegate this to the outside world, but that is definitely not the case here.